native-engine/datafusion-ext-plans/src/agg/spark_udaf_wrapper.rs

```diff
@@ -0,0 +1,344 @@
+        )?;
+        let batch_struct_array = StructArray::from(params_batch);
+        let mut export_ffi_batch_array = FFI_ArrowArray::new(&batch_struct_array.to_data());
+
+        // create zipped indices
+        let max_len = std::cmp::max(acc_idx.len(), partial_arg_idx.len());
+        let mut zipped_indices = Vec::with_capacity(max_len);
+        idx_for_zipped! {
+            ((acc_idx, updating_acc_idx) in (acc_idx, partial_arg_idx)) => {
+                zipped_indices.push((acc_idx as i64) << 32 | updating_acc_idx as i64);
+            }
+        }
+        let zipped_indices_array = jni_new_prim_array!(long, &zipped_indices[..])?;
+
+        jni_call!(SparkUDAFWrapperContext(self.jcontext()?.as_obj()).update(
+            accs.obj.as_obj(),
+            &mut export_ffi_batch_array as *mut FFI_ArrowArray as i64,
+            zipped_indices_array.as_obj(),
+        )-> ())
+    }
+
+    fn partial_merge(
+        &self,
+        accs: &mut AccColumnRef,
+        acc_idx: IdxSelection<'_>,
+        merging_accs: &mut AccColumnRef,
+        merging_acc_idx: IdxSelection<'_>,
+    ) -> Result<()> {
+        let accs = downcast_any!(accs, mut AccUnsafeRowsColumn).unwrap();
+        let merging_accs = downcast_any!(merging_accs, mut AccUnsafeRowsColumn).unwrap();
+
+        // create zipped indices
+        let max_len = std::cmp::max(acc_idx.len(), merging_acc_idx.len());
+        let mut zipped_indices = Vec::with_capacity(max_len);
+        idx_for_zipped! {
+            ((acc_idx, updating_acc_idx) in (acc_idx, merging_acc_idx)) => {
+                zipped_indices.push((acc_idx as i64) << 32 | updating_acc_idx as i64);
+            }
+        }
+        let zipped_indices_array = jni_new_prim_array!(long, &zipped_indices[..])?;
+
+        jni_call!(SparkUDAFWrapperContext(self.jcontext()?.as_obj()).merge(
+            accs.obj.as_obj(),
+            merging_accs.obj.as_obj(),
+            zipped_indices_array.as_obj(),
+        )-> ())
+    }
+
+    fn final_merge(&self, accs: &mut AccColumnRef, acc_idx: IdxSelection<'_>) -> Result<ArrayRef> {
+        let accs = downcast_any!(accs, mut AccUnsafeRowsColumn).unwrap();
+        let acc_indices = acc_idx.to_int32_vec();
+
+        let acc_idx_array = jni_new_prim_array!(int, &acc_indices[..])?;
+        let mut import_ffi_array = FFI_ArrowArray::empty();
+
+        jni_call!(SparkUDAFWrapperContext(self.jcontext()?.as_obj()).eval(
+            accs.obj.as_obj(),
+            acc_idx_array.as_obj(),
+            &mut import_ffi_array as *mut FFI_ArrowArray as i64,
+        )-> ())?;
+
+        // import output from context
+        let import_ffi_schema = FFI_ArrowSchema::try_from(self.import_schema.as_ref())?;
+        let import_struct_array =
+            make_array(unsafe { from_ffi(import_ffi_array, &import_ffi_schema)? });
+        let import_array = as_struct_array(&import_struct_array).column(0).clone();
+        Ok(import_array)
+    }
+}
+
+struct AccUnsafeRowsColumn {
```

Comment:
i would like a better name like `AccUDAFBufferRowsColumn` because it is not only for UnsafeRows.